### ISSUE TITLE
Full-size inline image from screenshot tool

### ISSUE THREAD

**Issue by raisingw** [qTox/qTox#5337]
Would it be possible, if it is known that the image is coming from the screenshot tool rather than a standard file send, for the client to display the image as wide as the current chat window (with the option to zoom to native size), similar to how uTox does it? uTox also does this with pasted images.

It would also be handy for this to not force the receiving user to accept it as a file that must be saved. The receiving end shows the image in the chat window at full size with an option to save but only do it this way if it's a qTox-created screenshot or pasted image - not one selected from a file open dialog.

**Comment by ilovelinux** [comment]
You're asking to store the image in RAM to show it without save on disk and then you can choose to save it?

It is pretty useless because you (obviously) download the image anyway and if you don't save it, you'll have to ask him to resend the file.

So what's the reason why you want this feature? Maybe I'm missing some use-cases

**Comment by raisingw** [comment]
That's exactly how uTox does it. My friends and I share a lot of images with each other (artwork) that we don't actually save. Some we do, some we don't, the option is nice. Unfortunately uTox is a crashy mess for me and doesn't get the dev activity that qTox obviously does. We'd like to switch but the inline image functionality of uTox keeps us with it.

Don't most other chat/messenger services allow people to post images into the chat without forcing the other person or group to save it to disk?

**Comment by anon596992** [comment]
I agree. In my opinion this is an important feature. Most if not all major instant messaging platforms support this, as it is very common for people to share images in conversations without the assumption of saving them.

Having to click and save the image every time and then hover over it or view it elsewhere is inconvenient for most people, which is why I think this is one of the features needed for Tox to become more popular.

I also think the image should show inline for every image sent, not only images sent using the qTox screenshot tool. Ideally, it would be an option that can be enabled or disabled in the settings, and maybe on a friend by friend basis like in the case of automatic file transfers.

Since there is already code for the preview of downloaded images ( https://github.com/qTox/qTox/blob/master/src/chatlog/content/filetransferwidget.cpp ), this should not be too hard to implement. The preview could simply be moved from the file transfer widget to the chatbox.

The other part is probably simply making images transfer automatically (if the option mentioned earlier is enabled) without having to enable automatic file transfers for every file, and then storing them in memory (or a temporary folder that is treated the same as the user's history). The only problem seems to be making sure it is actually an image being sent and not some other type of file regardless of its extension. uTox seems to already do this for its inline images: https://github.com/uTox/uTox/blob/develop/src/file_transfers.c

This is at least what I think would be ideal, but making the screenshot tool and pasted images work uTox-style as raisingw mentioned would already increase usability a lot.

**Comment by skeletonium** [comment]
I have created mockups of how this feature could look like (for more versions, feature ideas, and explanations, see below):

![MockupV1](https://user-images.githubusercontent.com/65868451/83981063-fcd77580-a909-11ea-88f5-477d5852935d.png)

Compared to uTox:

![uToxSide](https://user-images.githubusercontent.com/65868451/83981150-e7af1680-a90a-11ea-8d33-571f9ac46d19.png)

uTox has a small gap between the image and the file transfer bar, maybe that would look better:

![MockupV2](https://user-images.githubusercontent.com/65868451/83981148-e382f900-a90a-11ea-93a0-28e0565951c0.png)

Notice that, just like uTox, the bar says "Click to save", meaning the image would remain in memory until explicitly saved (same thing as uTox does). Alternatively, images could be considered part of history and stored in a subdirectory of the qTox database, and encrypted with the qTox profile password.

So maybe the bar shouldn't be green until the image is saved:

![MockupV2Gray](https://user-images.githubusercontent.com/65868451/83981238-a703cd00-a90b-11ea-8d53-0a5b64a121d4.png)

Also had some ideas for little features which could increase user experience for inline images, which would be more than what uTox does:

![MockupV3](https://user-images.githubusercontent.com/65868451/83981293-0e218180-a90c-11ea-9425-578becfe8493.png)

In summary, a button for rolling up and hiding the inline image, in case it is a distraction to the message box, and a cross button to discard the image completely. Note: I omitted the roll-up button for the saved image in the mockup, but in case this feature is implemented, there is probably no reason to not have it for saved images as well.

Here is how it would look with the image rolled up:
![MockupV3RolledUp](https://user-images.githubusercontent.com/65868451/83981383-ce0ece80-a90c-11ea-87e4-eb769ab99f9a.png)

And finally, here is an example of sending a message from qTox as well, in this case from the qTox screenshot tool:
![MockupV3WithSentImage](https://user-images.githubusercontent.com/65868451/83981399-eb439d00-a90c-11ea-8f21-984886d3a1bd.png)

Of course, while inline images would be convenient and comfortable for new users switching to qTox from popular instant messaging platforms, not everyone may want the feature, which is why an option should be offered in the settings, and customization for contacts:
![SettingsMockupA](https://user-images.githubusercontent.com/65868451/83981472-a704cc80-a90d-11ea-9426-a8af2ef30741.png)
![FriendMockupA](https://user-images.githubusercontent.com/65868451/83981475-acfaad80-a90d-11ea-8d5b-f73f210678cf.png)
In summary, as long as only "Accept images automatically" is checked, they would be stored in memory or, as mentioned earlier, as part of the history. "Save images automatically" would be the same as "Accept files automatically", but only for images.

Personally, I'm not sure if there is much of a need for it, but if we need something more fine-grained, there could also be a setting to only accept qTox and uTox screenshots as the original issue mentions. Maybe it's a bit of unnecessary complexity, but could be done if needed:
![SettingsMockupB](https://user-images.githubusercontent.com/65868451/83981523-301c0380-a90e-11ea-80ce-887ccd142ff6.png)
![FriendMockupB](https://user-images.githubusercontent.com/65868451/83981526-37dba800-a90e-11ea-8443-b362b3a49d39.png)

**Comment by skeletonium** [comment]
Continuation:

I was hoping to try implementing the feature myself, but since I have no C++ experience, I would likely get nowhere. But I will post here what I looked through in my experimentation in hopes it will be useful in implementing it:
Existing code for the preview of images, at line 512 in
https://github.com/qTox/qTox/blob/master/src/chatlog/content/filetransferwidget.cpp
Seems images are already treated selectively.

image.cpp, but not sure what uses the code here and the exact meaning of Q_UNUSED at lines 49, 50, 55.
https://github.com/qTox/qTox/blob/master/src/chatlog/content/image.cpp

uTox's handling of inline images, beginning at line 531 likely most relevant:
https://github.com/uTox/uTox/blob/develop/src/file_transfers.c

And then at line 822. Seems the automatic accepting of inline images is filename-based (uTox only has inline screenshots), and with some adjustments to the code, any valid .png would be accepted (confirmed with testing).

Header for uTox's image handling. Not sure if that would be of any use since qTox already has previews (which, with my limited knowledge, would make me assume handling of the images themselves is complete enough for the inline image feature).
https://github.com/uTox/uTox/blob/develop/src/native/image.h

**Comment by sudden6** [comment]
I like the UI ideas of @skeletonium for showing the image as large as possible in the chat log.

About the automatic downloading, I think we should couple this to the `auto-accept files` option. We should rename it to `auto-accept files and images` to make it more clear and if it is checked, downloaded images are shown in the chatlog automatically. If it is disabled, only explicitly downloaded images will be shown in big.

The reason I don't want another option is, that we already have a lot of options and that you can not reliably separate file downloads from image downloads, because a user might just rename their `evil.exe` to `evil.jpg` or do some file header trickery. Also we have a good system in place with the auto-accept save directory which can be set on a per friend basis to keep things sorted.

I don't want to store images in the message database either, because an SQL database is not very good at storing binary data, so it's going to bloat the database a lot.

**Comment by skeletonium** [comment]
@sudden6
I was thinking the option would be more for the convenience of not having to save every image, as original issue mentions
> It would also be handy for this to not force the receiving user to accept it as a file that must be saved.

> Don't most other chat/messenger services allow people to post images into the chat without forcing the other person or group to save it to disk?

But you have a good point about adding too many options. Maybe we could only have a single one for accepting the images to keep things simple (something like `Accept inline images`/`Show inline images automatically` for user friendliness), keeping them in memory like uTox does, and leave the rest (saving them) to either explicit user action or `Auto-accept files` option?

>  you can not reliably separate file downloads from image downloads, because a user might just rename their evil.exe to evil.jpg or do some file header trickery

In this case, I suppose the file would be discarded, being invalid as an image file (which is, I think, what uTox does), or appear as a normal file transfer prompt.

Edit: Tested with uTox, renaming a random file to utox-inline.png. It simply gets discarded with no prompt on the uTox side, as the code indicates:
https://github.com/uTox/uTox/blob/develop/src/file_transfers.c (Line 1001)
